added the download model from url feature

codezeros opened the pull request:
- support all downloadable urls 

**Models can be downloaded simply by copying and pasting their URLs.**

codezeros commented:
![preview](https://github.com/lllyasviel/Fooocus/assets/156826608/9a39fe07-e777-4cc9-b9fe-cc2b8ae68e85)

mashb1t commented on modules/download_models.py:
```diff
@@ -0,0 +1,31 @@
+    except Exception as e:
+        success_message = f"Download failed! please check url if it is correct."
+
+    return success_message
```
This can also be not a success message. Please rename to message.

mashb1t commented on modules/load_paths.py:
```diff
@@ -0,0 +1,96 @@
+
+    return folder_data
+
+def save_to_json(data, json_file):
+    """
+    This function takes data and a json file as input and writes the data to the json file.
+
+    'data' is the data to be written to the json file.
+
+    'json_file' is the path to the json file.
+
+    The data is written to the json file with an indentation of 4 spaces.
+    """
+    with open(json_file, 'w') as f:
+        json.dump(data, f, indent=4)
```
This function is only called once. Remove function and use code directly.

mashb1t commented on modules/load_paths.py:
```diff
@@ -0,0 +1,96 @@
+        print("Error: The specified folder does not exist.")
+        return
+
+    folder_data = get_folders_and_paths(root_folder)
```
keep in mind that all model folders can also be located at a different path than {project_root}/models.
Use all config paths here.

- path_checkpoints
- path_loras
- path_embeddings
- path_vae_approx
- path_upscale_models
- path_inpaint
- path_controlnet
- path_clip_vision
- path_fooocus_expansion

codezeros replied:
yes, now path can be changed in config json file also if we use the path directly from config.py the structure of the code base may lead to circular import error

mashb1t commented on modules/load_paths.py:
```diff
@@ -0,0 +1,96 @@
+    for item in items_to_delete:
+        del folder_data[item]
+
+    json_file_name = "model_config_path.json"
```
This is sufficient on application boot, no separate file is needed.

mashb1t commented on modules/download_models.py:
```diff
@@ -0,0 +1,31 @@
+
+    The function then tries to download the file from the URL and save it to the path. If the download is successful, a success message is returned. If the download fails, an error message is returned.
+    """
+    model_paths = read_model_config_path("./model_config_path.json")
```
read from config here

mashb1t commented on modules/shared_module.py:
```diff
@@ -0,0 +1,14 @@
+# shared_module.py
+import json
+
+def read_model_config_path(json_file_path):
```
not needed, use config here

mashb1t commented on .gitignore:
```diff
@@ -51,4 +51,6 @@ user_path_config-deprecated.txt
 /package-lock.json
 /.coverage*
 /auth.json
+/model_config_path.json
```
not needed

mashb1t commented on modules/download_models.py:
```diff
@@ -0,0 +1,31 @@
+from modules.model_loader import load_file_from_url
+from modules.shared_module import read_model_config_path
+
+def download_models(url, selected, file_name=None):
```
this can be migrated into modules/util.py

mashb1t commented on modules/download_models.py:
```diff
@@ -0,0 +1,31 @@
+    try:
+
+        load_file_from_url(url, model_dir=path, progress=True, file_name=file_name)
+        success_message = f"Download successful! Model saved to {path}."
```
add all output messages and tab names etc. to translation template to `en.json` and check with arg `--language en` if they're correct

mashb1t commented on webui.py:
```diff
@@ -110,18 +111,24 @@ def generate_clicked(task):
                 with gr.Column(scale=17):
-                    prompt = gr.Textbox(show_label=False, placeholder="Type prompt here or paste parameters.", elem_id='positive_prompt',
+                    prompt = gr.Textbox(show_label=False, placeholder="Type prompt here or paste parameters.",
+                                        elem_id='positive_prompt',
```
Please revert all autoformatting of the code so your PR only has changes of your feature. Code cleanup is done separately.

codezeros replied:
yes, the formatting is done automatically by ide automatically and now it has been fixed 👍

mashb1t commented on webui.py:
```diff
@@ -331,12 +379,73 @@ def update_history_link():
                     lambda: None, _js='()=>{refresh_style_localization();}')
+            with gr.Tab(label='download'):
+                """
+                This creates a new tab in the Gradio interface with the label 'download'.
```
Remove this comment and all following unnecessary comments, `gr.Tab(label='download')` is very much readable on its own without comment.
Add comments when they explain complexity, but not for trivial things which can be read.

mashb1t commented on webui.py:
```diff
@@ -331,12 +379,73 @@ def update_history_link():
+                """
+                with gr.Group():
+                    """
+                    This creates a new group in the 'download' tab. A group is a container for other Gradio components.
```
"A group is a container for other Gradio components." lol, thanks for the explanation ^^
 If you use ChatGPT or similar, make sure to refactor the code afterwards to be less verbose in comments.

codezeros replied:
just using copilot to simplify and speed up writing doc strings :)

mashb1t commented on webui.py:
```diff
@@ -331,12 +379,73 @@ def update_history_link():
+                        file_name = gr.Textbox(label="Enter File Name:", placeholder="Enter the File Name "
+                                                                                     "With its Extension.(Optional)")
+                        """
+                        This creates a new row in the group and adds a textbox to it. The textbox is labeled 'Enter File Name:' and is used for the user to input the name of the file to be downloaded. If the user does not provide a file name, the original file name from the URL will be used.
```
as mentioned...

mashb1t reviewed (changes requested):
General feedback: thank you for the implementation, this can be very useful if people don't know the file structure.

Code feedback:
- do not spam code comments for trivial things, as mentioned.
- do not autoformat whole files on existing projects, only your part of the implementation.
- add translations for every outpout, every message, etc. to en.json
- add argument for disabling the whole tab as not everyone wants/needs/allows users to download newc models, such as on hosted environments, as this may cause a huge increase in traffic (=> more costs) and may be abused, especially in a multi-user environment.

Looking forward to your adjustments.

codezeros commented:
> General feedback: thank you for the implementation, this can be very useful if people don't know the file structure.
> 
> Code feedback:
> 
>     * do not spam code comments for trivial things, as mentioned.
> 
>     * do not autoformat whole files on existing projects, only your part of the implementation.
> 
>     * add translations for every outpout, every message, etc. to en.json
> 
>     * add argument for disabling the whole tab as not everyone wants/needs/allows users to download newc models, such as on hosted environments, as this may cause a huge increase in traffic (=> more costs) and may be abused, especially in a multi-user environment.
> 
> 
> Looking forward to your adjustments.

@mashb1t thank you for your feedback 
i have now removed unnecessary comments and make the codebase simple and effective and we can set the path now directly in json config file instead of dynamically generate json on start of app each time run.

mashb1t commented on modules/download_models.py:
```diff
@@ -0,0 +1,21 @@
+
+
+def download_models(url, selected, file_name=None):
+    with open('./model_config_path.json', 'r') as f:
+        model_paths = json.load(f)
```
the content of this file must be configurable using the config settings. Please remove model_config_path.json and use config settings instead as combined array here.
If you run into circular imports then you'll have to rearrange the code accordingly.

mashb1t commented on modules/download_models.py:
```diff
@@ -0,0 +1,21 @@
+
+    try:
+        load_file_from_url(url, model_dir=path, progress=True, file_name=file_name)
+        message = f"Download successful! Model saved to {path}."
+    except Exception as e:
+        message = f"Download failed! Please check the URL and try again."
```
Please keep the translations (en.json) in mind for all outputs to the frontend, incl., messages, tab titles, labels, element names, etc.
Thanks!

mashb1t commented on modules/model_loader.py:
```diff
@@ -20,7 +20,12 @@ def load_file_from_url(
-        download_url_to_file(url, cached_file, progress=progress)
-    return cached_file
+        try:
+
```
```suggestion
```

mashb1t commented on modules/model_loader.py:
```diff
@@ -20,7 +20,12 @@ def load_file_from_url(
+            from torch.hub import download_url_to_file
+            download_url_to_file(url, cached_file, progress=progress)
+        except Exception as e:
+            print(f"Failed to download {url} to {cached_file}: {e}")
```
inconsistency in quotes, "{url}" used above but here it's only {url}. Please use "{url}".

```suggestion
            print(f"Failed to download "{url}" to {cached_file}, reason: {e}")
```

mashb1t commented on webui.py:
```diff
@@ -332,6 +333,32 @@ def update_history_link():
+
+                    choices = list(model_paths.keys())
+                    with gr.Row():
+                        url_input = gr.Textbox(label="Enter URL:")
```
```suggestion
                        url_input = gr.Textbox(label="URL")
```

mashb1t commented on webui.py:
```diff
@@ -332,6 +333,32 @@ def update_history_link():
+                        url_input = gr.Textbox(label="Enter URL:")
+                    with gr.Row():
+                        file_name = gr.Textbox(label="Enter File Name:", placeholder="File Name"
+                                                                                     "With its Extension.(Optional)")
```
Dangling string, please fix formatting here

mashb1t commented on webui.py:
```diff
@@ -332,6 +333,32 @@ def update_history_link():
+                    with gr.Row():
+                        url_input = gr.Textbox(label="Enter URL:")
+                    with gr.Row():
+                        file_name = gr.Textbox(label="Enter File Name:", placeholder="File Name"
```
```suggestion
                        file_name = gr.Textbox(label="File Name"
```

mashb1t commented on webui.py:
```diff
@@ -332,6 +333,32 @@ def update_history_link():
                     lambda: None, _js='()=>{refresh_style_localization();}')
 
+
+            with gr.Tab(label='download'):
```
```suggestion
            with gr.Tab(label='Download'):
```

mashb1t reviewed (changes requested):
keep in mind to add the translations to en.json as well as add an argument to completely not render the output for all elements when set.
Argument can be `--disable-download-tab` or similar. You can orient yourself on `--disable-metadata` + all occurrences of args_manager.args.disable_metadata in the code

uptotec commented:
hi! if you can only download from civitAI you can get the filename and folder from the api. civitai provide array of files the first one is always the model file. the api also give you the model type you can select the folder based on the model type.

codezeros commented:
> hi! if you can only download from civitAI you can get the filename and folder from the api. civitai provide array of files the first one is always the model file. the api also give you the model type you can select the folder based on the model type.

@uptotec yes, for that we need to add store feature using civitai api but this is for downloading models directly into the application using url only also i have tested with huggingface and civitai urls and some other it works! without any issues but note that some model may have sign-in access restriction for those we can not download directly just with url alone

codezeros commented:
i have modified according to your suggestion and also we can disable download tab using argument and also using checkbox in debug(still in progress any suggestions would help!) and added translation :)

mashb1t commented:
@codezeros thanks for the update, much appreciated. Will do the review during this week.
I'm wondering, which benefit would it have to be able to toggle the tab in the developer debug settings?

codezeros commented:
> @codezeros thanks for the update, much appreciated. Will do the review during this week. I'm wondering, which benefit would it have to be able to toggle the tab in the developer debug settings?


![image](https://github.com/lllyasviel/Fooocus/assets/156826608/cad3c186-df99-4cea-a505-ec8cb0eeeb6b)

@mashb1t  i thought user can also disable the tab in fooocus itself to simply make the config easier for non-tech users which may help

mashb1t commented on webui.py:
```diff
@@ -433,6 +453,7 @@ def update_history_link():
                         disable_seed_increment = gr.Checkbox(label='Disable seed increment',
                                                              info='Disable automatic seed increment when image number is > 1.',
                                                              value=False)
+                        disable_download_checkbox = gr.Checkbox(label='Disable Download Tab',info='Disable the download tab when clicked',value=modules.config.default_download_tab_checkbox,elem_classes='min_check')
```
set interactive to false when arg is set

mashb1t commented on webui.py:
```diff
@@ -541,6 +562,9 @@ def model_refresh_clicked():
         advanced_checkbox.change(lambda x: gr.update(visible=x), advanced_checkbox, advanced_column,
                                  queue=False, show_progress=False) \
             .then(fn=lambda: None, _js='refresh_grid_delayed', queue=False, show_progress=False)
+        disable_download_checkbox.change(lambda x: gr.update(visible=x), disable_download_checkbox, download_tab) \
```
do not process/output when arg is true

mashb1t commented on webui.py:
```diff
@@ -332,6 +333,25 @@ def update_history_link():
                     lambda: None, _js='()=>{refresh_style_localization();}')
 
+            if not args_manager.args.disable_download_tab:
+                    with gr.Tab(label='Download',visible=modules.config.default_download_tab_checkbox)as download_tab:
```
```suggestion
                    with gr.Tab(label='Download', visible=modules.config.default_download_tab_checkbox) as download_tab:
```

mashb1t commented on modules/config.py:
```diff
@@ -166,16 +166,45 @@ def get_dir_or_set_default(key, default_value, as_array=False, make_directory=Fa
-path_controlnet = get_dir_or_set_default('path_controlnet', '../models/controlnet/')
-path_clip_vision = get_dir_or_set_default('path_clip_vision', '../models/clip_vision/')
-path_fooocus_expansion = get_dir_or_set_default('path_fooocus_expansion', '../models/prompt_expansion/fooocus_expansion')
+config_paths = {
```
please do it the other way round.
Instead of using an array and then assigning variables with keys it's more robust to assign variables first and then build an array from vars. This eliminates the need for array key string reading.

mashb1t commented on modules/config.py:
```diff
@@ -318,6 +347,11 @@ def get_config_item_or_set_default(key, default_value, validator, disable_empty_
     default_value=False,
     validator=lambda x: isinstance(x, bool)
 )
+default_download_tab_checkbox = get_config_item_or_set_default(
```
please use a more descriptive naming here, e.g. `default_show_download_tab`

xhoxye commented:
Some personal advice:
1. You can list the built-in model download links in focus and add other backup download links. Select Download through the drop-down option to solve the problem of model download failure when focus is started.
2. For Civitai, you can refresh the update of the model version

mashb1t commented on modules/model_loader.py:
```diff
@@ -20,7 +20,11 @@ def load_file_from_url(
+            from torch.hub import download_url_to_file
+            download_url_to_file(url, cached_file, progress=progress)
+        except Exception as e:
+            print(f"Failed to download \"{url}\" to {cached_file}, reason : {e}")
```
@codezeros downloading an invalid URL still reports as successful to the frontend despite raising an exception.
This has to be fixed (reverted).

`Failed to download "invalid url" to /Volumes/data/private/development/ai/fooocus/models/checkpoints/invalid url, reason : unknown url type: 'invalid url'
`

![Screenshot 2024-03-09 at 16 19 39](https://github.com/lllyasviel/Fooocus/assets/9307310/0b08a872-5cc0-44bd-84a5-0d19a53bfae6)

codezeros replied:
@mashb1t  thanks for notifying the problem i will look into it and also making some improvements based on previous feedback (in progress)

Alexdnk commented:
Maybe make the download feature not in a separate tab, but right below in the "Model" tab using gradio accordion something that:
 `gr.Accordion('Downloads'):`?

sd-jack commented:
That's enhancement is great! Thank you @codezeros

mashb1t commented:
closing as stale
